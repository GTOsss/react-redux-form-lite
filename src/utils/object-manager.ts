import stringToPath from 'lodash.topath';

/**
 *
 * @param {object} state Object will be expand
 * @param {string} path Path in object
 * @param {*} value Value will be add
 * @param {number} pathIndex Not require
 * @returns {object} State
 */
export const setIn = (state, path, value, pathIndex = 0) => {
  const pathArray = pathIndex === 0 ? stringToPath(path) : path;

  if (pathIndex >= pathArray.length) {
    return value;
  }

  const first = pathArray[pathIndex];
  const firstState = state && (Array.isArray(state) ? state[Number(first)] : state[first]);
  const next = setIn(firstState, pathArray, value, pathIndex + 1);

  if (!state) {
    if (isNaN(first)) {
      return { [first]: next };
    }
    const initialized = [];
    // @ts-ignore
    initialized[parseInt(first, 10)] = next;
    return initialized;
  }

  if (Array.isArray(state)) {
    // @ts-ignore
    const copy = [].concat(state);
    copy[parseInt(first, 10)] = next;
    return copy;
  }

  return {
    ...state,
    [first]: next,
  };
};

/**
 *
 * @param {object} state Redux state
 * @param {string} field Path to field
 * @param {*?} defaultValue
 * @returns {object} State
 */
export const getIn = (state, field, defaultValue?) => {
  if (!state) {
    return defaultValue;
  }

  const path = stringToPath(field);
  const { length } = path;
  if (!length) {
    return defaultValue;
  }

  let result = state;
  for (let i = 0; i < length && result; i += 1) {
    result = result[path[i]];
  }

  if (!result) {
    return defaultValue;
  }

  return result;
};

/**
 * @param {object} state Redux state
 * @param path
 * @param removeEmpty
 * @param index
 * @returns {object} State
 */
const deleteInByPath = (state, path: Array<string>, removeEmpty: boolean = false, index: number = 0) => {
  const currentKey = path[index];

  if (!state && (state !== 0)) {
    return state;
  }

  const isEndPoint = (path.length - 1) === index;
  const isArray = Array.isArray(state);
  const isObject = !isArray && (typeof state === 'object');
  const newState = isArray ? [...state] : {...state};

  if (isObject) {
    if (isEndPoint) {
      // tslint:disable-next-line:no-dynamic-delete
      delete newState[currentKey];
      return newState;
    } else {
      const result = deleteInByPath(newState[currentKey], path, removeEmpty, index + 1);
      const isRemoveEmpty = !result || (removeEmpty && (!result || !Object.keys(result).length));
      if (isRemoveEmpty) {
        // tslint:disable-next-line:no-dynamic-delete
        delete newState[currentKey];
      } else {
        newState[currentKey] = result;
      }
      return newState;
    }
  } else if (isArray) {
    if (isEndPoint) {
      return newState.filter((el, i) => Number(currentKey) !== i);
    } else {
      const result = deleteInByPath(newState[currentKey], path, removeEmpty, index + 1);
      const isRemoveEmpty = !result || (removeEmpty && (!result || !Object.keys(result).length));
      if (isRemoveEmpty) {
        return newState.filter((el, i) => Number(currentKey) !== i);
      } else {
        newState[currentKey] = result;
      }
      return newState;
    }
  }
};

export const deleteIn = (state, field, removeEmpty: boolean = false) => {
  const path = stringToPath(field);
  return deleteInByPath(state, path, removeEmpty);
};

const isObject = (item: any): boolean =>
  item && (typeof item === 'object') && !Array.isArray(item);

/**
 * Deep merge two objects.
 * @param target
 * @param sources
 */
export const mergeDeep = (target, ...sources) => {
  if (!sources.length) {
    return target;
  }
  const source = sources.shift();

  if (isObject(target) && isObject(source)) {
    Object.keys(source).forEach((key) => {
      if (isObject(source[key])) {
        if (!target[key]) {
          Object.assign(target, { [key]: {} });
        }
        mergeDeep(target[key], source[key]);
      } else {
        Object.assign(target, { [key]: source[key] });
      }
    });
  }

  return mergeDeep(target, ...sources);
};
