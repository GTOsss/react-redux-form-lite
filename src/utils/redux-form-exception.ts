const UNKNOWN = 'UNKNOWN';

export default class ReduxFormException {
  type: string;
  description: string;

  constructor(type: string) {
    this.type = type;

    switch (type) {
      default: {
        this.type = UNKNOWN;
        this.description = 'Unknown error.';
        break;
      }
    }
  }
}
