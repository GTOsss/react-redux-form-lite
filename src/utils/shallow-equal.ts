const is = (x, y) => {
  // SameValue algorithm
  if (x === y) {
    // Steps 1-5, 7-10
    // Steps 6.b-6.e: +0 != -0
    // Added the nonzero y check to make Flow happy, but it is redundant
    return x !== 0 || y !== 0 || 1 / x === 1 / y;
  } else {
    // Step 6.a: NaN == NaN
    return x !== x && y !== y;
  }
};

export const shallowEqual = (objA, objB) => {
  const hasOwnProperty = Object.prototype.hasOwnProperty;

  if (is(objA, objB)) {
    return true;
  }

  if (typeof objA !== 'object' || objA === null || typeof objB !== 'object' || objB === null) {
    return false;
  }

  const keysA = Object.keys(objA);
  const keysB = Object.keys(objB);

  if (keysA.length !== keysB.length) {
    return false;
  }

  // Test for A's keys different from B.
  for (const el of keysA) {
    if (!hasOwnProperty.call(objB, el) || !is(objA[el], objB[el])) {
      return false;
    }
  }

  return true;
};