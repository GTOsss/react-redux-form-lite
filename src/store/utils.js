import {addToObjectByPath, deleteIn, getIn} from '../utils/object-manager';

/**
 *
 * @param state
 * @param form
 * @param map
 * @param {'error' | 'warning'} type
 */
const updateMessages = (state, form, map, type) => {
  const pathMap = {
    error: {
      meta: (key) => `${form}.meta.${key}.error`,
      messagesMap: `${form}.form.errorsMap`,
      hasMessages: `${form}.form.hasErrors`,
    },
    warning: {
      meta: (key) => `${form}.meta.${key}.warning`,
      messagesMap: `${form}.form.warningsMap`,
      hasMessages: `${form}.form.hasWarnings`,
    }
  };

  const messagesMapDefault = getIn(state, pathMap[type].messagesMap, {});
  const currentMessagesMap = {...messagesMapDefault, ...map};
  let newState = addToObjectByPath(state, pathMap[type].messagesMap, currentMessagesMap);

  Object.entries(currentMessagesMap).forEach(([key, value]) => {
    newState = addToObjectByPath(newState, pathMap[type].meta(key), value || '');
    if (!(value || (value === 0))) {
      newState = deleteIn(newState, `${pathMap[type].messagesMap}.${key}`);
      newState = addToObjectByPath(newState, `${pathMap[type].meta(key)}`, '');
    }
  });

  const messagesMap = getIn(newState, pathMap[type].messagesMap);
  const hasMessages = Object.keys(messagesMap).length !== 0;

  return addToObjectByPath(newState, pathMap[type].hasMessages, hasMessages);
};

export const updateErrors = (state, form, map) => updateMessages(state, form, map, 'error');

export const updateWarnings = (state, form, map) => updateMessages(state, form, map, 'warning');

/**
 * @param state
 * @param form
 * @param {object?} map
 * @param {object?} map.errors
 * @param {object?} map.warnings
 */
export const updateErrorsAndWarnings = (state, form, {errors = {}, warnings = {}}) => {
  let newState = updateErrors(state, form, errors);
  return updateWarnings(newState, form, warnings);
};