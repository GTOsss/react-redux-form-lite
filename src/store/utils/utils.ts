import {setIn, deleteIn, getIn} from '../../utils/object-manager';
import {
  IMapValidateErrorsAndWarnings,
  IValues,
  IMapSubmitValidate,
} from '../types';

/**
 *
 * @param {object} state State
 * @param {string} form Name of form
 * @param {object} map Map with errors or warnings
 * @param {boolean} submitted Submitted
 * @param {'error' | 'warning'} type Type of messages
 * @param {string?} wizard Wizard key
 * @return {object} Updated state
 */
const updateMessages = (
  state: IFullReduxFormState<any>,
  form: string,
  map: MapMessages<any>,
  type: 'error' | 'warning',
  submitted?: boolean,
  wizard?: string,
): IFullReduxFormState<any> => {
  const pathMap = {
    error: {
      meta: (key) => `${form}.meta.${key}.error`,
      messagesMap: `${form}.form.errorsMap`,
      hasMessages: `${form}.form.hasErrors`,
      wizardHasMessages: `${wizard}.wizard.hasErrors`,
      wizardMessagesMap: `${wizard}.wizard.errorsMap`,
    },
    warning: {
      meta: (key) => `${form}.meta.${key}.warning`,
      messagesMap: `${form}.form.warningsMap`,
      hasMessages: `${form}.form.hasWarnings`,
      wizardHasMessages: `${wizard}.wizard.hasWarnings`,
      wizardMessagesMap: `${wizard}.wizard.warningsMap`,
    },
  };

  const messagesMapDefault = getIn(state, pathMap[type].messagesMap, {});
  const currentMessagesMap = {...messagesMapDefault, ...map};
  let newState = setIn(state, pathMap[type].messagesMap, currentMessagesMap);
  if (typeof submitted === 'boolean') {
    newState = setIn(newState, `${form}.form.submitted`, submitted);
  }

  Object.entries(currentMessagesMap).forEach(([key, value]) => {
    const isRegisteredField = Boolean(getIn(state, `${form}.meta.${key}`));
    if (isRegisteredField) {
      newState = setIn(newState, pathMap[type].meta(key), value || '');
    }
    if (isRegisteredField && !(value || (value === 0))) {
      newState = deleteIn(newState, `${pathMap[type].messagesMap}.${key}`);
      newState = setIn(newState, `${pathMap[type].meta(key)}`, '');
    }
    if (!isRegisteredField) {
      newState = deleteIn(newState, `${pathMap[type].messagesMap}.${key}`);
    }
  });

  const messagesMap = getIn(newState, pathMap[type].messagesMap);
  const hasMessages = Object.keys(messagesMap).length !== 0;

  if (wizard) {
    newState = setIn(newState, pathMap[type].wizardHasMessages, hasMessages);
    newState = setIn(newState, pathMap[type].wizardMessagesMap, messagesMap);
  }

  return setIn(newState, pathMap[type].hasMessages, hasMessages);
};

export const updateErrors = (
  state: IFullReduxFormState<any>,
  form: string,
  map: MapMessages<any>,
  submitted?: boolean,
  wizard?: string,
): IFullReduxFormState<any> =>
  updateMessages(state, form, map, 'error', submitted, wizard);

export const updateWarnings = (
  state: IFullReduxFormState<any>,
  form: string,
  map: MapMessages<any>,
  submitted?: boolean,
  wizard?: string,
): IFullReduxFormState<any> =>
  updateMessages(state, form, map, 'warning', submitted, wizard);

/**
 * @param {object} state State
 * @param {string} form Name of form
 * @param {object?} map Map with errors and warnings
 * @param {object?} map.errors Map errors
 * @param {object?} map.warnings Map warnings
 * @param {boolean?} submitted Submitted
 * @param {string?} wizard Key of wizard
 * @return {object} Updated state
 */
export const updateErrorsAndWarnings = (
  state: IFullReduxFormState<any>,
  form: string,
  {
    errors = {},
    warnings = {},
  }: IMapErrorsAndWarningsMessages<any>,
  submitted?: boolean,
  wizard?: string,
): IFullReduxFormState<any> => {
  const newState = updateErrors(state, form, errors, submitted, wizard);
  return updateWarnings(newState, form, warnings, submitted, wizard);
};

const getMessageMap = (
  key: string,
  value: any,
  validate: ValidateProps,
  targetMap: MapMessages<any>,
): void => {
  if (validate && (typeof validate === 'function')) {
    targetMap[key] = validate(value);
  } else if (validate && Array.isArray(validate)) {
    for (let i = 0; i < validate.length; i += 1) {
      const result = validate[i](value);
      if (result) {
        targetMap[key] = result;
        break;
      }

      if (i === (validate.length - 1)) {
        targetMap[key] = undefined;
      }
    }
  }
};

export const validateFormByValues = (
  values: IValues,
  validateMap: IMapValidateErrorsAndWarnings = {},
  submitValidateMap: IMapSubmitValidate = {},
) => {
  let resultErrors = {};
  let resultWarnings = {};

  Object.entries(values).forEach(([key, value]) => {
    if (validateMap.validate) {
      getMessageMap(key, value, validateMap.validate[key], resultErrors);
    }
    if (validateMap.warn) {
      getMessageMap(key, value, validateMap.warn[key], resultWarnings);
    }
  });

  if (submitValidateMap && typeof submitValidateMap.validate === 'function') {
    resultErrors = {...resultErrors, ...submitValidateMap.validate(values)};
  }

  if (submitValidateMap && typeof submitValidateMap.warn === 'function') {
    resultWarnings = {...resultWarnings, ...submitValidateMap.warn(values)};
  }

  return {errors: resultErrors, warnings: resultWarnings};
};

interface IValidateFormByStateParams {
  state: IFullReduxFormState<any>;
  form: string;
  validateMap: IMapValidateErrorsAndWarnings;
  submitValidateMap: IMapSubmitValidate;
  submitted?: boolean;
  wizard?: string;
}

/**
 * @param {object} param params
 * @param {object} param.state state
 * @param {string} param.form from name
 * @param {object} param.validateMap field level validate map with validate methods
 * @param {object} param.submitValidateMap submit level validate map with validate methods
 * @param {boolean?} param.submitted true if need set submitted
 * @returns {*|{}|Object} state or errorAndWarningMap
 */
export const validateFormByState = ({
  state,
  form,
  validateMap,
  submitValidateMap,
  submitted,
  wizard,
}: IValidateFormByStateParams): IFullReduxFormState<any> => {
  const formState = state[form];
  const {values = {}} = formState || {};
  const result = validateFormByValues(values, validateMap, submitValidateMap);
  return updateErrorsAndWarnings(state, form, result, submitted, wizard);
};
