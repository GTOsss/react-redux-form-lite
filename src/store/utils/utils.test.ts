import {updateWarnings, updateErrors, updateErrorsAndWarnings, mergeMessages} from './utils';

interface IMockTwoFieldsValues {
  testField1: string;
  testField2: string;
}

const defaultFormState = {
  errorsMap: {},
  warningsMap: {},
  hasErrors: false,
  hasWarnings: false,
  activeField: '',
  blurred: false,
  changed: false,
  focused: false,
  submitted: false,
};

const defaultFieldMetaState = {
  warning: '',
  error: '',
  active: false,
  blurred: false,
  changed: false,
  focused: false,
};

const getMockState = (): IFullReduxFormState<IMockTwoFieldsValues> => ({
  testForm: {
    form: {
      ...defaultFormState,
    },
    meta: {
      testField1: {
        ...defaultFieldMetaState,
      },
      testField2: {
        ...defaultFieldMetaState,
      },
    },
    values: {
      testField1: '',
      testField2: '',
    },
  },
});

describe('Store utils.', () => {
  test('updateErrors field1', () => {
    const form = 'testForm';
    const mockState = getMockState();

    const map = {testField1: 'Error for field 1'};
    expect(updateErrors(mockState, form, map)).toMatchSnapshot();
  });

  test('updateErrors field2', () => {
    const form = 'testForm';
    const mockState = getMockState();

    const map = {testField2: 'Error for field 2'};
    expect(updateErrors(mockState, form, map)).toMatchSnapshot();
  });

  test('updateErrors field1, field2', () => {
    const form = 'testForm';
    const mockState = getMockState();

    const map = {
      testField1: 'Error for field 1',
      testField2: 'Error for field 2',
    };
    expect(updateErrors(mockState, form, map)).toMatchSnapshot();
  });

  test('updateErrors (update to undefined)', () => {
    const form = 'testForm';
    const mockState = getMockState();

    const map = {
      testField1: 'Error1',
      testField2: 'Error2',
    };
    let result = updateErrors(mockState, form, map);

    // remove testField1
    let resultUndefined = updateErrors(result, form, {testField1: undefined});
    let expectState: IFullReduxFormState<IMockTwoFieldsValues> = {
      testForm: {
        form: {
          ...defaultFormState,
          errorsMap: {testField2: 'Error2'},
          hasErrors: true,
        },
        meta: {
          testField1: {
            ...defaultFieldMetaState,
          },
          testField2: {
            ...defaultFieldMetaState,
            error: 'Error2',
          },
        },
        values: {
          testField1: '',
          testField2: '',
        },
      },
    };
    expect(resultUndefined).toEqual(expectState);

    let resultNull = updateErrors(result, form, {testField1: null});
    let resultEmptyString = updateErrors(result, form, {testField1: ''});
    expect(resultUndefined).toEqual(resultNull);
    expect(resultUndefined).toEqual(resultEmptyString);

    // remove testField2
    resultUndefined = updateErrors(resultUndefined, form, {testField2: undefined});
    result = resultUndefined;
    expectState = {
      testForm: {
        form: {
          ...defaultFormState,
        },
        meta: {
          testField1: {
            ...defaultFieldMetaState,
          },
          testField2: {
            ...defaultFieldMetaState,
          },
        },
        values: {
          testField1: '',
          testField2: '',
        },
      },
    };
    expect(resultUndefined).toEqual(expectState);

    resultNull = updateErrors(result, form, {testField2: null});
    resultEmptyString = updateErrors(result, form, {testField2: ''});
    expect(resultUndefined).toEqual(resultNull);
    expect(resultUndefined).toEqual(resultEmptyString);
  });

  test('updateWarnings (update to undefined)', () => {
    const form = 'testForm';
    const mockState = getMockState();

    const map = {
      testField1: 'Warning1',
      testField2: 'Warning2',
    };
    let result = updateWarnings(mockState, form, map);

    // remove testField1
    let resultUndefined = updateWarnings(result, form, {testField1: undefined});
    let expectState: IFullReduxFormState<IMockTwoFieldsValues> = {
      testForm: {
        form: {
          ...defaultFormState,
          warningsMap: {testField2: 'Warning2'},
          hasWarnings: true,
        },
        meta: {
          testField1: {
            ...defaultFieldMetaState,
          },
          testField2: {
            ...defaultFieldMetaState,
            warning: 'Warning2',
          },
        },
        values: {
          testField1: '',
          testField2: '',
        },
      },
    };

    expect(resultUndefined).toEqual(expectState);

    let resultNull = updateWarnings(result, form, {testField1: null});
    let resultEmptyString = updateWarnings(result, form, {testField1: ''});
    expect(resultUndefined).toEqual(resultNull);
    expect(resultUndefined).toEqual(resultEmptyString);

    // remove testField2
    resultUndefined = updateWarnings(resultUndefined, form, {testField2: undefined});
    result = resultUndefined;
    expectState = {
      testForm: {
        form: {
          ...defaultFormState,
        },
        meta: {
          testField1: {
            ...defaultFieldMetaState,
          },
          testField2: {
            ...defaultFieldMetaState,
          },
        },
        values: {
          testField1: '',
          testField2: '',
        },
      },
    };
    expect(resultUndefined).toEqual(expectState);

    resultNull = updateErrors(result, form, {testField2: null});
    resultEmptyString = updateErrors(result, form, {testField2: ''});
    expect(resultUndefined).toEqual(resultNull);
    expect(resultUndefined).toEqual(resultEmptyString);
  });

  test('updateWarnings field1', () => {
    const form = 'testForm';
    const mockState = getMockState();

    const map = {testField1: 'Warning for field 1'};
    expect(updateWarnings(mockState, form, map)).toMatchSnapshot();
  });

  test('updateWarnings field2', () => {
    const form = 'testForm';
    const mockState = getMockState();

    const map = {testField2: 'Warning for field 2'};
    expect(updateWarnings(mockState, form, map)).toMatchSnapshot();
  });

  test('updateWarning field1, field2', () => {
    const form = 'testForm';
    const mockState = getMockState();

    const map = {
      testField1: 'Warning for field 1',
      testField2: 'Warning for field 2',
    };
    expect(updateWarnings(mockState, form, map)).toMatchSnapshot();
  });

  test('updateErrorsAndWarnings field1(error)', () => {
    const form = 'testForm';
    const mockState = getMockState();

    const map = {
      errors: {testField1: 'Error for field 1'},
    };
    expect(updateErrorsAndWarnings(mockState, form, map)).toMatchSnapshot();
  });

  test('updateErrorsAndWarnings field1(warning)', () => {
    const form = 'testForm';
    const mockState = getMockState();

    const map = {
      warnings: {testField1: 'Warning for field 1'},
    };
    expect(updateErrorsAndWarnings(mockState, form, map)).toMatchSnapshot();
  });

  test('updateErrorsAndWarnings field1(error), field1(warning)', () => {
    const form = 'testForm';
    const mockState = getMockState();

    const map3 = {
      errors: {testField1: 'Error for field 1'},
      warnings: {testField1: 'Warning for field 1'},
    };
    expect(updateErrorsAndWarnings(mockState, form, map3)).toMatchSnapshot();
  });

  test('updateErrorsAndWarnings field1(error, warning) field2(error, warning)', () => {
    const form = 'testForm';
    const mockState = getMockState();

    const map4 = {
      errors: {testField1: 'Error for field 1', testField2: 'Error for field 2'},
      warnings: {testField1: 'Warning for field 2', testField2: 'Warning for field 2'},
    };
    expect(updateErrorsAndWarnings(mockState, form, map4)).toMatchSnapshot();
  });

  test('updateWarnings field1 and then field2', () => {
    const form = 'testForm';
    let mockState = getMockState();

    const map1 = {testField1: 'Warning for field 1'};
    const map2 = {testField2: 'Warning for field 2'};
    mockState = updateWarnings(mockState, form, map1);
    // tslint:disable-next-line:no-console
    console.log(mockState);
    expect(updateWarnings(mockState, form, map2)).toMatchSnapshot();
  });

  test('updateWarnings field1 and then field2 = undefined', () => {
    const form = 'testForm';
    let mockState = getMockState();

    const map1 = {testField1: 'Warning for field 1'};
    const map2 = {testField2: undefined};
    mockState = updateWarnings(mockState, form, map1);
    // tslint:disable-next-line:no-console
    console.log(mockState);
    expect(updateWarnings(mockState, form, map2)).toMatchSnapshot();
  });

  test('mergeMessages 1', () => {
    const objectA = {
      field1: '1',
      field2: '2',
    };

    const objectB = {
      field1: undefined,
    };

    expect(mergeMessages(objectA, objectB)).toStrictEqual({field1: '1', field2: '2'});
    expect(mergeMessages(objectB, objectA)).toStrictEqual({field1: '1', field2: '2'});
  });

  test('mergeMessages 2', () => {
    const objectA = {
      field1: '1',
      field2: '2',
    };

    const objectB = {
      field1: undefined,
      field3: '3',
    };

    const result = {...objectA};

    expect(mergeMessages(objectA, objectB)).toStrictEqual({field1: '1', field2: '2', field3: '3'});
    expect(mergeMessages(objectB, objectA)).toStrictEqual({field1: '1', field2: '2', field3: '3'});
  });
});