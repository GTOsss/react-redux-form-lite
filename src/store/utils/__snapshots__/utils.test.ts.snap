// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store utils. updateErrors field1 1`] = `
Object {
  "testForm": Object {
    "form": Object {
      "activeField": "",
      "blurred": false,
      "changed": false,
      "errorsMap": Object {
        "testField1": "Error for field 1",
      },
      "focused": false,
      "hasErrors": true,
      "hasWarnings": false,
      "submitted": false,
      "warningsMap": Object {},
    },
    "meta": Object {
      "testField1": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "Error for field 1",
        "focused": false,
        "warning": "",
      },
      "testField2": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "",
        "focused": false,
        "warning": "",
      },
    },
    "values": Object {
      "testField1": "",
      "testField2": "",
    },
  },
}
`;

exports[`Store utils. updateErrors field1, field2 1`] = `
Object {
  "testForm": Object {
    "form": Object {
      "activeField": "",
      "blurred": false,
      "changed": false,
      "errorsMap": Object {
        "testField1": "Error for field 1",
        "testField2": "Error for field 2",
      },
      "focused": false,
      "hasErrors": true,
      "hasWarnings": false,
      "submitted": false,
      "warningsMap": Object {},
    },
    "meta": Object {
      "testField1": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "Error for field 1",
        "focused": false,
        "warning": "",
      },
      "testField2": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "Error for field 2",
        "focused": false,
        "warning": "",
      },
    },
    "values": Object {
      "testField1": "",
      "testField2": "",
    },
  },
}
`;

exports[`Store utils. updateErrors field2 1`] = `
Object {
  "testForm": Object {
    "form": Object {
      "activeField": "",
      "blurred": false,
      "changed": false,
      "errorsMap": Object {
        "testField2": "Error for field 2",
      },
      "focused": false,
      "hasErrors": true,
      "hasWarnings": false,
      "submitted": false,
      "warningsMap": Object {},
    },
    "meta": Object {
      "testField1": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "",
        "focused": false,
        "warning": "",
      },
      "testField2": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "Error for field 2",
        "focused": false,
        "warning": "",
      },
    },
    "values": Object {
      "testField1": "",
      "testField2": "",
    },
  },
}
`;

exports[`Store utils. updateErrorsAndWarnings field1(error) 1`] = `
Object {
  "testForm": Object {
    "form": Object {
      "activeField": "",
      "blurred": false,
      "changed": false,
      "errorsMap": Object {
        "testField1": "Error for field 1",
      },
      "focused": false,
      "hasErrors": true,
      "hasWarnings": false,
      "submitted": false,
      "warningsMap": Object {},
    },
    "meta": Object {
      "testField1": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "Error for field 1",
        "focused": false,
        "warning": "",
      },
      "testField2": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "",
        "focused": false,
        "warning": "",
      },
    },
    "values": Object {
      "testField1": "",
      "testField2": "",
    },
  },
}
`;

exports[`Store utils. updateErrorsAndWarnings field1(error), field1(warning) 1`] = `
Object {
  "testForm": Object {
    "form": Object {
      "activeField": "",
      "blurred": false,
      "changed": false,
      "errorsMap": Object {
        "testField1": "Error for field 1",
      },
      "focused": false,
      "hasErrors": true,
      "hasWarnings": true,
      "submitted": false,
      "warningsMap": Object {
        "testField1": "Warning for field 1",
      },
    },
    "meta": Object {
      "testField1": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "Error for field 1",
        "focused": false,
        "warning": "Warning for field 1",
      },
      "testField2": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "",
        "focused": false,
        "warning": "",
      },
    },
    "values": Object {
      "testField1": "",
      "testField2": "",
    },
  },
}
`;

exports[`Store utils. updateErrorsAndWarnings field1(error, warning) field2(error, warning) 1`] = `
Object {
  "testForm": Object {
    "form": Object {
      "activeField": "",
      "blurred": false,
      "changed": false,
      "errorsMap": Object {
        "testField1": "Error for field 1",
        "testField2": "Error for field 2",
      },
      "focused": false,
      "hasErrors": true,
      "hasWarnings": true,
      "submitted": false,
      "warningsMap": Object {
        "testField1": "Warning for field 2",
        "testField2": "Warning for field 2",
      },
    },
    "meta": Object {
      "testField1": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "Error for field 1",
        "focused": false,
        "warning": "Warning for field 2",
      },
      "testField2": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "Error for field 2",
        "focused": false,
        "warning": "Warning for field 2",
      },
    },
    "values": Object {
      "testField1": "",
      "testField2": "",
    },
  },
}
`;

exports[`Store utils. updateErrorsAndWarnings field1(warning) 1`] = `
Object {
  "testForm": Object {
    "form": Object {
      "activeField": "",
      "blurred": false,
      "changed": false,
      "errorsMap": Object {},
      "focused": false,
      "hasErrors": false,
      "hasWarnings": true,
      "submitted": false,
      "warningsMap": Object {
        "testField1": "Warning for field 1",
      },
    },
    "meta": Object {
      "testField1": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "",
        "focused": false,
        "warning": "Warning for field 1",
      },
      "testField2": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "",
        "focused": false,
        "warning": "",
      },
    },
    "values": Object {
      "testField1": "",
      "testField2": "",
    },
  },
}
`;

exports[`Store utils. updateWarning field1, field2 1`] = `
Object {
  "testForm": Object {
    "form": Object {
      "activeField": "",
      "blurred": false,
      "changed": false,
      "errorsMap": Object {},
      "focused": false,
      "hasErrors": false,
      "hasWarnings": true,
      "submitted": false,
      "warningsMap": Object {
        "testField1": "Warning for field 1",
        "testField2": "Warning for field 2",
      },
    },
    "meta": Object {
      "testField1": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "",
        "focused": false,
        "warning": "Warning for field 1",
      },
      "testField2": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "",
        "focused": false,
        "warning": "Warning for field 2",
      },
    },
    "values": Object {
      "testField1": "",
      "testField2": "",
    },
  },
}
`;

exports[`Store utils. updateWarnings field1 1`] = `
Object {
  "testForm": Object {
    "form": Object {
      "activeField": "",
      "blurred": false,
      "changed": false,
      "errorsMap": Object {},
      "focused": false,
      "hasErrors": false,
      "hasWarnings": true,
      "submitted": false,
      "warningsMap": Object {
        "testField1": "Warning for field 1",
      },
    },
    "meta": Object {
      "testField1": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "",
        "focused": false,
        "warning": "Warning for field 1",
      },
      "testField2": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "",
        "focused": false,
        "warning": "",
      },
    },
    "values": Object {
      "testField1": "",
      "testField2": "",
    },
  },
}
`;

exports[`Store utils. updateWarnings field1 and then field2 = undefined 1`] = `
Object {
  "testForm": Object {
    "form": Object {
      "activeField": "",
      "blurred": false,
      "changed": false,
      "errorsMap": Object {},
      "focused": false,
      "hasErrors": false,
      "hasWarnings": true,
      "submitted": false,
      "warningsMap": Object {
        "testField1": "Warning for field 1",
      },
    },
    "meta": Object {
      "testField1": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "",
        "focused": false,
        "warning": "Warning for field 1",
      },
      "testField2": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "",
        "focused": false,
        "warning": "",
      },
    },
    "values": Object {
      "testField1": "",
      "testField2": "",
    },
  },
}
`;

exports[`Store utils. updateWarnings field1 and then field2 1`] = `
Object {
  "testForm": Object {
    "form": Object {
      "activeField": "",
      "blurred": false,
      "changed": false,
      "errorsMap": Object {},
      "focused": false,
      "hasErrors": false,
      "hasWarnings": true,
      "submitted": false,
      "warningsMap": Object {
        "testField1": "Warning for field 1",
        "testField2": "Warning for field 2",
      },
    },
    "meta": Object {
      "testField1": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "",
        "focused": false,
        "warning": "Warning for field 1",
      },
      "testField2": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "",
        "focused": false,
        "warning": "Warning for field 2",
      },
    },
    "values": Object {
      "testField1": "",
      "testField2": "",
    },
  },
}
`;

exports[`Store utils. updateWarnings field2 1`] = `
Object {
  "testForm": Object {
    "form": Object {
      "activeField": "",
      "blurred": false,
      "changed": false,
      "errorsMap": Object {},
      "focused": false,
      "hasErrors": false,
      "hasWarnings": true,
      "submitted": false,
      "warningsMap": Object {
        "testField2": "Warning for field 2",
      },
    },
    "meta": Object {
      "testField1": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "",
        "focused": false,
        "warning": "",
      },
      "testField2": Object {
        "active": false,
        "blurred": false,
        "changed": false,
        "error": "",
        "focused": false,
        "warning": "Warning for field 2",
      },
    },
    "values": Object {
      "testField1": "",
      "testField2": "",
    },
  },
}
`;
